File: celltypist_annotate/critic.py
"""
celltypist_annotate critic — Evaluates cell type annotation quality

FEEDBACK KEYS (must match parameter_science_guide conditions in skill.yaml):
    high_unassigned       → p_thres too high or model mismatch
    low_confidence_labels → mean probability too low
    fragmented_voting     → min_prop too high; many mixed clusters unassigned
    over_homogenized      → too few cell types (possible over-annotation)
"""
import numpy as np


def critic_post_process(adata, context=None):
    context = context or {}
    protocol = context.get('protocol', '10x Genomics')
    tissue   = context.get('tissue', 'immune')

    thresholds = {
        '10x Genomics': {'max_pct_unassigned': 20.0, 'min_mean_prob': 0.55, 'min_cell_types': 2},
        'Smart-seq2':   {'max_pct_unassigned': 15.0, 'min_mean_prob': 0.60, 'min_cell_types': 2},
        'Drop-seq':     {'max_pct_unassigned': 25.0, 'min_mean_prob': 0.50, 'min_cell_types': 2},
        'CITE-seq':     {'max_pct_unassigned': 20.0, 'min_mean_prob': 0.55, 'min_cell_types': 2},
    }
    t = thresholds.get(protocol, thresholds['10x Genomics'])

    metrics = {}
    warnings = []
    feedback = []

    # Find label column
    label_col = None
    for col in ['majority_voting', 'predicted_labels']:
        if col in adata.obs.columns:
            label_col = col
            break
    if label_col is None:
        return {
            'metrics': {},
            'warnings': ['CRITICAL: No CellTypist annotation columns found in adata.obs'],
            'feedback': [],
            'success': False,
            'context_used': protocol
        }

    labels = adata.obs[label_col]
    metrics['n_cells'] = int(len(labels))
    metrics['n_cell_types'] = int(labels.nunique())
    metrics['pct_unassigned'] = float((labels == 'Unassigned').sum() / len(labels) * 100)

    # Confidence score
    for conf_col in ['conf_score', 'max_prob']:
        if conf_col in adata.obs.columns:
            metrics['mean_max_probability'] = float(adata.obs[conf_col].mean())
            break
    else:
        metrics['mean_max_probability'] = float('nan')

    # Top cell types
    metrics['top3_celltypes'] = (
        labels[labels != 'Unassigned'].value_counts().head(3).index.tolist()
    )

    # Evaluate
    if metrics['pct_unassigned'] > t['max_pct_unassigned']:
        warnings.append(
            f"HIGH UNASSIGNED: {metrics['pct_unassigned']:.1f}% cells unassigned "
            f"(max {t['max_pct_unassigned']:.0f}% for {protocol}). "
            "Consider lowering p_thres or using a more appropriate model."
        )
        feedback.append('high_unassigned')

    if (not np.isnan(metrics['mean_max_probability']) and
            metrics['mean_max_probability'] < t['min_mean_prob']):
        warnings.append(
            f"LOW CONFIDENCE: mean probability {metrics['mean_max_probability']:.3f} "
            f"< threshold {t['min_mean_prob']}. "
            "Model may not match the tissue type. Check celltypist.models.models_description()."
        )
        feedback.append('low_confidence_labels')

    if metrics['n_cell_types'] < t['min_cell_types']:
        warnings.append(
            f"TOO FEW CELL TYPES: only {metrics['n_cell_types']} type(s) detected. "
            "Check if majority_voting is over-homogenizing results."
        )
        feedback.append('over_homogenized')

    return {
        'metrics': metrics,
        'warnings': warnings,
        'feedback': feedback,
        'success': not bool(feedback),
        'context_used': protocol
    }

File: singler_classifycells/critic.R
#' singler_classifycells critic — Evaluates SingleR annotation quality
#'
#' FEEDBACK KEYS (match parameter_science_guide in skill.yaml):
#'   too_few_cell_types  → labels_field too coarse; try label.fine
#'   too_many_pruned     → high ambiguity; switch to label.main
#'   high_na_rate        → prune=TRUE removing too many cells
#'   low_confidence_labels → mean delta.next too low

critic_post_process <- function(obj, context = NULL) {
    if (is.null(context)) context <- list()
    protocol <- if (!is.null(context$protocol)) context$protocol else "10x Genomics"

    thresholds <- list(
        "10x Genomics" = list(max_pct_pruned = 20.0, min_delta_next = 0.05, min_cell_types = 2L),
        "Smart-seq2"   = list(max_pct_pruned = 15.0, min_delta_next = 0.08, min_cell_types = 2L),
        "Drop-seq"     = list(max_pct_pruned = 25.0, min_delta_next = 0.04, min_cell_types = 2L),
        "CITE-seq"     = list(max_pct_pruned = 20.0, min_delta_next = 0.05, min_cell_types = 2L)
    )
    t <- if (!is.null(thresholds[[protocol]])) thresholds[[protocol]] else thresholds[["10x Genomics"]]

    metrics  <- list()
    warnings <- character(0)
    feedback <- character(0)

    # Find SingleR label column
    is_seurat <- inherits(obj, "Seurat")
    meta <- if (is_seurat) obj@meta.data else as.data.frame(SummarizedExperiment::colData(obj))
    singler_cols <- grep("^SingleR\\.", colnames(meta), value = TRUE)

    if (length(singler_cols) == 0L) {
        return(list(
            metrics      = list(),
            warnings     = "CRITICAL: No SingleR annotation columns found — run SingleR first",
            feedback     = character(0),
            success      = FALSE,
            context_used = protocol
        ))
    }

    label_col <- singler_cols[!grepl("score|delta", singler_cols)][1]
    labels <- meta[[label_col]]

    metrics$n_cells      <- as.integer(nrow(meta))
    metrics$n_cell_types <- as.integer(length(unique(na.omit(labels))))
    metrics$pct_pruned   <- mean(is.na(labels)) * 100

    if ("SingleR.delta.next" %in% colnames(meta)) {
        metrics$mean_delta_next <- mean(meta$SingleR.delta.next, na.rm = TRUE)
    } else {
        metrics$mean_delta_next <- NA_real_
    }

    metrics$top3_celltypes <- names(
        sort(table(na.omit(labels)), decreasing = TRUE)
    )[1:min(3L, metrics$n_cell_types)]

    if (metrics$pct_pruned > t$max_pct_pruned) {
        warnings <- c(warnings, sprintf(
            "HIGH PRUNING RATE: %.1f%% cells pruned (max %.0f%% for %s). "
            "Consider switching to 'label.main' for broader categories.",
            metrics$pct_pruned, t$max_pct_pruned, protocol
        ))
        feedback <- c(feedback, "too_many_pruned")
    }

    if (!is.na(metrics$mean_delta_next) && metrics$mean_delta_next < t$min_delta_next) {
        warnings <- c(warnings, sprintf(
            "LOW CONFIDENCE: mean delta.next = %.3f (threshold %.3f). "
            "Many cells are ambiguous between closely related types.",
            metrics$mean_delta_next, t$min_delta_next
        ))
        feedback <- c(feedback, "low_confidence_labels")
    }

    if (metrics$n_cell_types < t$min_cell_types) {
        warnings <- c(warnings, sprintf(
            "TOO FEW CELL TYPES: %d type(s) detected. "
            "Try 'label.fine' for more granular annotation.",
            metrics$n_cell_types
        ))
        feedback <- c(feedback, "too_few_cell_types")
    }

    list(
        metrics      = metrics,
        warnings     = warnings,
        feedback     = feedback,
        success      = length(feedback) == 0L,
        context_used = protocol
    )
}
